sudoku_board.cc

```diff
@@ -1,123 +1,181 @@
-    if (get_cell_options(r, c)==std::set<int>({})){
-        if (get_board_values()[r][c]==EMPTY_CELL_SIGN){
-            return LOSE;
-        }
-        else {
-            return WIN;
-        }
-    }
-    return PROG;
+void SudokuBoard::set_cell_value(size_t r, size_t c, int val) {
+  if (get_cell_options(r, c).count(val) <= 0) {
+    throw std::invalid_argument(OPTION_INVALID_ERR_MSG);
+  }
+  m_board_values[r][c] = val;
+  update_possible_options();
 }
 
-std::set<int> SudokuBoard::get_cell_options(int r, int c){
-    return m_possible_options[r][c];
+// operators
+
+// TODO: do I need to add more functions (rule of 5)?
+SudokuBoard &SudokuBoard::operator=(const SudokuBoard &other) {
+  if (this->m_size != other.m_size) {
+    throw std::invalid_argument(SIZE_INCOMPATIBALE_ERR_MSG);
+  }
+  if (this != &other) {
+    m_possible_options = other.m_possible_options;
+    m_board_values = m_board_values;
+  }
+  return *this;
 }
 
-void SudokuBoard::update_cell_value(int r, int c, int val){
-    m_board_values[r][c] = val;
-    update_possible_options();
+// methods
+
+STATUS_OPTION SudokuBoard::calc_status() const {
+  bool is_win = true;
+  for (size_t i = 0; i < m_size; i++) {
+    for (size_t j = 0; j < m_size; j++) {
+      STATUS_OPTION cell_status = calc_cell_status(i, j);
+      if (cell_status == LOSE) {
+        return LOSE;
+      } else if (cell_status == PROG) {
+        is_win = false;
+      }
+    }
+  }
+  if (is_win) {
+    return WIN;
+  }
+  return PROG;
 }
 
-void SudokuBoard::update_possible_options(){
+STATUS_OPTION SudokuBoard::calc_cell_status(size_t r, size_t c) const {
+  if (get_cell_options(r, c).empty()) {
+    if (get_board_values()[r][c] == EMPTY_CELL_SIGN) {
+      return LOSE;
+    } else {
+      return WIN;
+    }
+  }
+  return PROG;
+}
 
+void SudokuBoard::update_possible_options() {
+  for (size_t i = 0; i < m_size; i++) {
+    for (size_t j = 0; j < m_size; j++) {
+      update_cell_possible_options(i, j);
+    }
+  }
 }
 
-// int main(int argc, char const *argv[])
-// {
-//     /* code */
-//     return 0;
-// }
+std::set<int>::iterator
+SudokuBoard::erase_option(size_t guess_cell_r, size_t guess_cell_c,
+                          std::set<int>::iterator option_iter) {
+  return m_possible_options[guess_cell_r][guess_cell_c].erase(option_iter);
+}
 
-// ------------------------------- private methods ---------------------------------
+bool SudokuBoard::is_empty_cell(size_t r, size_t c) const {
+  return m_board_values[r][c] == EMPTY_CELL_SIGN;
+}
 
-std::set<int> SudokuBoard::get_values_in_cube(int r, int c){
-    std::set<int> values_in_cube{};
-    int cube_top_r = r - (r % CUBE_SIZE);
-    int cube_left_c = c - (c % CUBE_SIZE);
-    for (int i = cube_top_r; i < cube_top_r + CUBE_SIZE; i++){
-        for (int j = cube_left_c; j < cube_left_c + CUBE_SIZE; j++){
-            values_in_cube.insert(m_board_values[i][j]);
-        }
+std::set<int> SudokuBoard::calc_forbidden_values_in_cube(size_t r,
```

Comment:
```suggestion
std::set<int> SudokuBoard::get_values_in_cube(size_t r,
```

Maybe you call it "forbidden", but these are just the current values in the cube...
If you don't think I'm right you are welcome to comment back and explain, this is true for all errors